# Null, CASE

Nous couvrirons les expressions **NULL** et **CASE** dans cette section. Une valeur **NULL** n'est pas une valeur, tout comme un **None** ou **NaN** en Python indique une valeur vide. L'expression **CASE** nous permet d'associer des conditions aux valeurs résultantes, un peu comme un **if/ elif else** en Python.


# SET UP

In [14]:
%pip install -q ipywidgets pandas matplotlib sqlite3 ipython-sql

In [2]:
import sqlite3
import pandas as pd



conn = sqlite3.connect('data/company_operations.db')

# Valeurs NULL

Jeter un coup d'oeil sur les rapports suivants : 'LJVE08D', 'EP4AKZR', '1FC27OH', 'F4DEAK3' se trouvant dans la table WEATHER_MONITOR


In [4]:
sql="""
SELECT * FROM WEATHER_MONITOR
WHERE REPORT_CODE IN ('LJVE08D', 'EP4AKZR', '1FC27OH', 'F4DEAK3')
"""
pd.read_sql(sql,conn)

ID REPORT_CODE REPORT_DATE  LOCATION_ID  TEMPERATURE  OVERCAST  RAIN  \
0    45     EP4AKZR  2021-03-11           32         64.5         0   NaN   
1    98     LJVE08D  2021-04-09           45         58.8         1   NaN   
2  1449     F4DEAK3  2021-01-15           17         35.3         1   0.0   
3  2967     1FC27OH  2020-12-01           16         89.7         0   NaN   

   SNOW  LIGHTNING  HAIL  TORNADO  
0   0.0          0     0        0  
1   0.0          0     0        0  
2   NaN          0     0        0  
3   0.0          0     0        0

* Donner tous les enregistrements tels que RAIN contient une valeur Null ( None )

| | ID | RAPPORT_CODE | RAPPORT_DATE | LOCATION_ID | TEMPERATURE | OVERCAST | RAIN | SNOW | LIGHTNING | HAIL | TORNADO |
|----:|-----:|:--------------|:--------------|--------------:|--------------:|-----------:|:-------|-------:|------------:|-------:|----------:|
| 0 | 9 | G0UINBG | 2021-05-04 | 14 | 62.2 | 1 |None | 0 | 0 | 0 | 0 |
| 1 | 17 | 89U7PF3 | 2021-05-02 | 2 | 67,8 | 0 |None | 0 | 0 | 0 | 0 |
| 2 | 45 | EP4AKZR | 2021-03-11 | 32 | 64,5 | 0 |None | 0 | 0 | 0 | 0 |
| 3 | 80 | EPQO1H8 | 2021-05-09 | 31 | 54.1 | 0 |None | 0 | 0 | 0 | 0 |
| 159 | 2967 | 1FC27OH | 2020-12-01 | 16 | 89,7 | 0 |None | 0 | 0 | 0 | 0 |
| 160 | 2973 | OXDUXW5 | 2020-11-02 | 44 | 100,2 | 1 |None | 0 | 0 | 0 | 0 |
| 161 | 2975 | EA8O3S0 | 2020-12-01 | 6 | 92.2 | 1 |None | 0 | 0 | 0 | 0 |





In [5]:
sql ="""
SELECT * FROM WEATHER_MONITOR
-- filtrer les null : IS NULL
WHERE RAIN IS NULL
"""
pd.read_sql(sql,conn).tail(3)

ID REPORT_CODE REPORT_DATE  LOCATION_ID  TEMPERATURE  OVERCAST  RAIN  \
159  2967     1FC27OH  2020-12-01           16         89.7         0  None   
160  2973     OXDUXW5  2020-11-02           44        100.2         1  None   
161  2975     EA8O3S0  2020-12-01            6         92.2         1  None   

     SNOW  LIGHTNING  HAIL  TORNADO  
159   0.0          0     0        0  
160   0.0          0     0        0  
161   0.0          0     0        0

* Filtrer les valeurs telsque RAIN est non NULL

In [6]:
sql ="""
SELECT * FROM WEATHER_MONITOR
-- filtrer les NON null : IS NOT NULL
WHERE RAIN IS NOT NULL
"""
pd.read_sql(sql,conn)

ID REPORT_CODE REPORT_DATE  LOCATION_ID  TEMPERATURE  OVERCAST  RAIN  \
0        1     UVYMMWW  2021-03-20            0         66.0         1  3.74   
1        2     7VVYE2L  2021-04-10           24         61.3         0  0.00   
2        3     PJVNOSP  2021-02-26           32         61.6         1  1.58   
3        4     3B19P7S  2021-05-30           39         66.3         0  0.00   
4        5     EHVUPGY  2021-04-09           48         58.5         0  0.00   
...    ...         ...         ...          ...          ...       ...   ...   
2833  2996     RWG4DY4  2020-11-26            6        106.6         1  0.00   
2834  2997     9A9EQNQ  2020-11-29           12         85.8         1  2.24   
2835  2998     C7ALVIK  2021-01-14           48         96.8         0  0.00   
2836  2999     V660AWC  2021-01-10           29        103.6         1  2.73   
2837  3000     IC35HCE  2021-01-07           48         64.0         1  2.32   

      SNOW  LIGHTNING  HAIL  TORNADO  
0      0.0          0     0        0  
1      0.0          0     0        0  
2      0.0          0     0        0  
3      0.0          0     0        0  
4      0.0          0     0        0  
...    ...        ...   ...      ...  
2833   0.0          0     0        0  
2834   0.0          0     0        0  
2835   0.0          0     0        0  
2836   0.0          0     0        0  
2837   0.0          0     0        0  

[2838 rows x 11 columns]

* Filtrer RAIN avec al conditions RAIN > 0 exclu automatiquement les valeurs NULL

In [7]:
sql ="""
SELECT * FROM WEATHER_MONITOR
WHERE RAIN > 0
"""
pd.read_sql(sql,conn)

ID REPORT_CODE REPORT_DATE  LOCATION_ID  TEMPERATURE  OVERCAST  RAIN  \
0       1     UVYMMWW  2021-03-20            0         66.0         1  3.74   
1       3     PJVNOSP  2021-02-26           32         61.6         1  1.58   
2       8     R238Q5U  2021-05-15           39         62.9         1  1.45   
3      14     DWGYB58  2021-05-01           17         60.0         1  2.92   
4      15     6ALC472  2021-05-22            1         56.2         1  2.45   
..    ...         ...         ...          ...          ...       ...   ...   
825  2991     B92HYWK  2020-11-30           17         98.3         1  0.91   
826  2995     XV1UQZ4  2021-01-15            6         89.8         1  1.69   
827  2997     9A9EQNQ  2020-11-29           12         85.8         1  2.24   
828  2999     V660AWC  2021-01-10           29        103.6         1  2.73   
829  3000     IC35HCE  2021-01-07           48         64.0         1  2.32   

     SNOW  LIGHTNING  HAIL  TORNADO  
0     0.0          0     0        0  
1     0.0          0     0        0  
2     0.0          0     0        0  
3     0.0          1     0        0  
4     0.0          0     0        0  
..    ...        ...   ...      ...  
825   0.0          0     0        0  
826   0.0          0     0        0  
827   0.0          0     0        0  
828   0.0          0     0        0  
829   0.0          0     0        0  

[830 rows x 11 columns]

* Filtrer les valurs de RAIN > 0 **OU** les valeurs NULL

In [8]:
sql ="""
SELECT * FROM WEATHER_MONITOR
WHERE RAIN > 0 OR RAIN IS NULL
"""
pd.read_sql(sql,conn).head(10)

ID REPORT_CODE REPORT_DATE  LOCATION_ID  TEMPERATURE  OVERCAST  RAIN  SNOW  \
0   1     UVYMMWW  2021-03-20            0         66.0         1  3.74   0.0   
1   3     PJVNOSP  2021-02-26           32         61.6         1  1.58   0.0   
2   8     R238Q5U  2021-05-15           39         62.9         1  1.45   0.0   
3   9     G0UINBG  2021-05-04           14         62.2         1   NaN   0.0   
4  14     DWGYB58  2021-05-01           17         60.0         1  2.92   0.0   
5  15     6ALC472  2021-05-22            1         56.2         1  2.45   0.0   
6  16     H3TOK27  2021-03-15           46         69.4         1  3.16   0.0   
7  17     89U7PF3  2021-05-02            2         67.8         0   NaN   0.0   
8  18     OBO0LTC  2021-02-09           47         59.5         1  3.00   0.0   
9  20     H9C85JV  2021-05-14            9         60.6         1  2.31   0.0   

   LIGHTNING  HAIL  TORNADO  
0          0     0        0  
1          0     0        0  
2          0     0        0  
3          0     0        0  
4          1     0        0  
5          0     0        0  
6          1     1        1  
7          0     0        0  
8          1     0        0  
9          0     0        0

# COALESCE()

Convertie la valeur Null en une une valeur définie par l'utilisateur
```sql
-- convertir toutes les valeurs qui sont Null dans la colonne RAIN par 0 , et laisser les autres valeurs tels quelles sont
COALESCE(RAIN,0)
```


In [9]:
sql ="""
SELECT RAIN,COALESCE(RAIN,0) AS COALESCE_  FROM WEATHER_MONITOR
"""
pd.read_sql(sql,conn).head(20)

RAIN  COALESCE_
0   3.74       3.74
1   0.00       0.00
2   1.58       1.58
3   0.00       0.00
4   0.00       0.00
5   0.00       0.00
6   0.00       0.00
7   1.45       1.45
8    NaN       0.00
9   0.00       0.00
10  0.00       0.00
11  0.00       0.00
12  0.00       0.00
13  2.92       2.92
14  2.45       2.45
15  3.16       3.16
16   NaN       0.00
17  3.00       3.00
18  0.00       0.00
19  2.31       2.31

In [10]:
sql = """
SELECT REPORT_CODE,
       RAIN,
       COALESCE(RAIN, -1) COALESCE_RAIN
FROM WEATHER_MONITOR
WHERE REPORT_CODE IN ('G0UINBG', 'PJVNOSP')
"""
pd.read_sql(sql,conn)

REPORT_CODE  RAIN  COALESCE_RAIN
0     PJVNOSP  1.58           1.58
1     G0UINBG   NaN          -1.00

# CASE

In [11]:
sql ="""
SELECT REPORT_CODE,
      TEMPERATURE
FROM WEATHER_MONITOR
"""
pd.read_sql(sql,conn)

REPORT_CODE  TEMPERATURE
0        UVYMMWW         66.0
1        7VVYE2L         61.3
2        PJVNOSP         61.6
3        3B19P7S         66.3
4        EHVUPGY         58.5
...          ...          ...
2995     RWG4DY4        106.6
2996     9A9EQNQ         85.8
2997     C7ALVIK         96.8
2998     V660AWC        103.6
2999     IC35HCE         64.0

[3000 rows x 2 columns]

* AFFICHER ceci
  - SI la température >=78 --> HOT
  - SI la température >=60 et < 78 --> MILD
  - SI la température < 60 --> COLD

-----------


|      | REPORT_CODE   |   TEMPERATURE | TEMPERATURE_LABEL   |
|-----:|:--------------|--------------:|:--------------------|
|    0 | UVYMMWW       |          66   | MILD                |
|    1 | 7VVYE2L       |          61.3 | MILD                |
|    2 | PJVNOSP       |          61.6 | MILD                |
|    3 | 3B19P7S       |          66.3 | MILD                |
|    4 | EHVUPGY       |          58.5 | COLD                |
|    5 | ZBAHLGO       |          65   | MILD                |

In [12]:
sql = """
SELECT REPORT_CODE,
      TEMPERATURE,
      CASE
        WHEN TEMPERATURE >=78 THEN 'HOT'
        WHEN TEMPERATURE >=60 THEN 'MILD'
        ELSE 'COLD'
      END AS TEMPERATURE_LABEL
FROM WEATHER_MONITOR
"""
pd.read_sql(sql,conn)

REPORT_CODE  TEMPERATURE TEMPERATURE_LABEL
0        UVYMMWW         66.0              MILD
1        7VVYE2L         61.3              MILD
2        PJVNOSP         61.6              MILD
3        3B19P7S         66.3              MILD
4        EHVUPGY         58.5              COLD
...          ...          ...               ...
2995     RWG4DY4        106.6               HOT
2996     9A9EQNQ         85.8               HOT
2997     C7ALVIK         96.8               HOT
2998     V660AWC        103.6               HOT
2999     IC35HCE         64.0              MILD

[3000 rows x 3 columns]

In [15]:
%load_ext sql

In [17]:
%sql sqlite:///data/company_operations.db

In [18]:
%%sql
SELECT REPORT_CODE,
      TEMPERATURE,
      CASE
        WHEN TEMPERATURE >=78 THEN 'HOT'
        WHEN TEMPERATURE >=60 THEN 'MILD'
        ELSE 'COLD'
      END AS TEMPERATURE_LABEL
FROM WEATHER_MONITOR
LIMIT 5

 * sqlite:///data/company_operations.db
Done.


[('UVYMMWW', 66.0, 'MILD'),
 ('7VVYE2L', 61.3, 'MILD'),
 ('PJVNOSP', 61.6, 'MILD'),
 ('3B19P7S', 66.3, 'MILD'),
 ('EHVUPGY', 58.5, 'COLD')]

```sql
SELECT REPORT_CODE,
      TEMPERATURE,
      CASE
        WHEN TEMPERATURE >=78 THEN 'HOT'
        WHEN TEMPERATURE >=60 THEN 'MILD'
        ELSE 'COLD'
      END AS TEMPERATURE_LABEL
FROM WEATHER_MONITOR
```

Dans le code SQL précédent le ELSE affecte 'COLD' aux valeurs de temperatures qui sont < 60 et aussi les valeur NULL

In [19]:
sql = """
SELECT REPORT_CODE,
      TEMPERATURE,
      CASE
        WHEN TEMPERATURE >=78 THEN 'HOT'
        WHEN TEMPERATURE >=60 THEN 'MILD'
        WHEN TEMPERATURE < 60 THEN 'COLD'
        -- ELSE traite seulement les valeurs Null
        ELSE 'N/A'
      END AS TEMPERATURE_LABEL
FROM WEATHER_MONITOR
"""
pd.read_sql(sql,conn)

REPORT_CODE  TEMPERATURE TEMPERATURE_LABEL
0        UVYMMWW         66.0              MILD
1        7VVYE2L         61.3              MILD
2        PJVNOSP         61.6              MILD
3        3B19P7S         66.3              MILD
4        EHVUPGY         58.5              COLD
...          ...          ...               ...
2995     RWG4DY4        106.6               HOT
2996     9A9EQNQ         85.8               HOT
2997     C7ALVIK         96.8               HOT
2998     V660AWC        103.6               HOT
2999     IC35HCE         64.0              MILD

[3000 rows x 3 columns]

In [20]:
df = pd.read_sql(sql,conn)

In [21]:
df[df['TEMPERATURE_LABEL'] == 'N/A']

REPORT_CODE  TEMPERATURE TEMPERATURE_LABEL
465     U91W7VE          NaN               N/A
616     RDEU2AW          NaN               N/A
997     5LC6JX2          NaN               N/A

# REMARQUES COALESCE() et CASE
* Remplacer `COALESCE(RAIN,0)` par `CASE`
```sql
SELECT RAIN,COALESCE(RAIN,0) AS COALESCE_  FROM WEATHER_MONITOR
```



In [22]:
sql ="""
SELECT RAIN,COALESCE(RAIN,0) AS COALESCE_  FROM WEATHER_MONITOR
"""
pd.read_sql(sql,conn).head(20)

RAIN  COALESCE_
0   3.74       3.74
1   0.00       0.00
2   1.58       1.58
3   0.00       0.00
4   0.00       0.00
5   0.00       0.00
6   0.00       0.00
7   1.45       1.45
8    NaN       0.00
9   0.00       0.00
10  0.00       0.00
11  0.00       0.00
12  0.00       0.00
13  2.92       2.92
14  2.45       2.45
15  3.16       3.16
16   NaN       0.00
17  3.00       3.00
18  0.00       0.00
19  2.31       2.31

In [23]:
sql = """
SELECT RAIN,
      CASE
          WHEN RAIN IS NULL THEN 1
          ELSE RAIN
      END AS COALESCE_
FROM WEATHER_MONITOR
"""
pd.read_sql(sql,conn).head(20)

RAIN  COALESCE_
0   3.74       3.74
1   0.00       0.00
2   1.58       1.58
3   0.00       0.00
4   0.00       0.00
5   0.00       0.00
6   0.00       0.00
7   1.45       1.45
8    NaN       1.00
9   0.00       0.00
10  0.00       0.00
11  0.00       0.00
12  0.00       0.00
13  2.92       2.92
14  2.45       2.45
15  3.16       3.16
16   NaN       1.00
17  3.00       3.00
18  0.00       0.00
19  2.31       2.31

# NULL CASE truc et astuces

* Calculer Total `Rain` groupé par `YEAR` et `MONTH` seulement pour l'année  2021 ( de la table `WEATHER_MONITOR`)

|    |   YEAR |   MONTH |   TOTAL_RAIN |
|---:|-------:|--------:|-------------:|
|  0 |   2021 |       1 |       316.27 |
|  1 |   2021 |       2 |       138.07 |
|  2 |   2021 |       3 |       129.03 |
|  3 |   2021 |       4 |       153.79 |
|  4 |   2021 |       5 |       158.24 |

In [24]:
%%sql
SELECT * FROM WEATHER_MONITOR
LIMIT 5

 * sqlite:///data/company_operations.db
Done.


[(1, 'UVYMMWW', '2021-03-20', 0, 66.0, 1, 3.74, 0.0, 0, 0, 0),
 (2, '7VVYE2L', '2021-04-10', 24, 61.3, 0, 0.0, 0.0, 0, 0, 0),
 (3, 'PJVNOSP', '2021-02-26', 32, 61.6, 1, 1.58, 0.0, 0, 0, 0),
 (4, '3B19P7S', '2021-05-30', 39, 66.3, 0, 0.0, 0.0, 0, 0, 0),
 (5, 'EHVUPGY', '2021-04-09', 48, 58.5, 0, 0.0, 0.0, 0, 0, 0)]

In [25]:
%%sql
SELECT CAST(strftime('%Y', REPORT_DATE) AS INTEGER) AS YEAR,
       CAST(strftime('%m', REPORT_DATE) AS INTEGER ) AS MONTH,
       SUM(RAIN) AS TOTAL_RAIN
FROM WEATHER_MONITOR
WHERE YEAR = 2021
GROUP BY YEAR, MONTH


 * sqlite:///data/company_operations.db
Done.


[(2021, 1, 316.2699999999998),
 (2021, 2, 138.07000000000002),
 (2021, 3, 129.03),
 (2021, 4, 153.78999999999996),
 (2021, 5, 158.23999999999998)]

In [26]:
sql = """
SELECT CAST(strftime('%Y', REPORT_DATE) AS INTEGER) AS YEAR,
       CAST(strftime('%m', REPORT_DATE) AS INTEGER ) AS MONTH,
       SUM(RAIN) AS TOTAL_RAIN
FROM WEATHER_MONITOR
WHERE YEAR = 2021
GROUP BY YEAR, MONTH
"""
pd.read_sql(sql,conn)

YEAR  MONTH  TOTAL_RAIN
0  2021      1      316.27
1  2021      2      138.07
2  2021      3      129.03
3  2021      4      153.79
4  2021      5      158.24

SQLITE

```sql
SELECT CAST(strftime('%Y', REPORT_DATE) AS INTEGER) AS YEAR,
       CAST(strftime('%m', REPORT_DATE) AS INTEGER ) AS MONTH,
       SUM(RAIN) AS TOTAL_RAIN
FROM WEATHER_MONITOR
WHERE YEAR = 2021
GROUP BY YEAR, MONTH
```



ORACLE SQL

```sql
SELECT
    TO_NUMBER(TO_CHAR(REPORT_DATE, 'YYYY')) AS YEAR,
    TO_NUMBER(TO_CHAR(REPORT_DATE, 'MM')) AS MONTH,
    SUM(RAIN) AS TOTAL_RAIN
FROM
    WEATHER_MONITOR
WHERE
    EXTRACT(YEAR FROM REPORT_DATE) = 2021
GROUP BY
    TO_NUMBER(TO_CHAR(REPORT_DATE, 'YYYY')),
    TO_NUMBER(TO_CHAR(REPORT_DATE, 'MM'));
```

SQL SERVER (T-SQL)

```sql
SELECT
    YEAR(REPORT_DATE) AS YEAR,
    MONTH(REPORT_DATE) AS MONTH,
    SUM(RAIN) AS TOTAL_RAIN
FROM
    WEATHER_MONITOR
WHERE
    YEAR(REPORT_DATE) = 2021
GROUP BY
    YEAR(REPORT_DATE), MONTH(REPORT_DATE);
```

* On veut diviser TOTAL_RAIN  en deux colonnes, une  quand  TORNADO = 1  et une autre  pour  TORNADO = 0. Quel est le problème ici ?

| | ANNÉE | MOIS | TOTAL_TORNADO_RAIN | TOTAL_NON_TORNADO_RAIN |
|---:|-------:|--------:|---------------------:|-------------------------:|
|  0 |   2021 |       2 |                15.22 |                    15.22 |
|  1 |   2021 |       3 |                24.92 |                    24.92 |
|  2 |   2021 |       4 |                 9.87 |                     9.87 |
|  3 |   2021 |       5 |                19.88 |                    19.88 |


In [27]:
%%sql
--by Kenza
SELECT CAST(strftime('%Y', REPORT_DATE) AS INTEGER) AS YEAR,
       CAST(strftime('%m', REPORT_DATE) AS INTEGER ) AS MONTH,
       tornado,
       round(SUM(RAIN),2) AS TOTAL_RAIN
FROM WEATHER_MONITOR
WHERE YEAR = 2021
GROUP BY YEAR, MONTH, tornado

 * sqlite:///data/company_operations.db
Done.


[(2021, 1, 0, 316.27),
 (2021, 2, 0, 122.85),
 (2021, 2, 1, 15.22),
 (2021, 3, 0, 104.11),
 (2021, 3, 1, 24.92),
 (2021, 4, 0, 143.92),
 (2021, 4, 1, 9.87),
 (2021, 5, 0, 138.36),
 (2021, 5, 1, 19.88)]

In [28]:
sql = """
---le filtre s'applique aux deux colonnes
SELECT CAST(strftime('%Y', REPORT_DATE) AS INTEGER) AS YEAR,
       CAST(strftime('%m', REPORT_DATE) AS INTEGER ) AS MONTH,
       SUM(RAIN) AS TOTALTORNADO_RAIN,
       SUM(RAIN) AS TOTAL_NON_TORNADO_RAIN
FROM WEATHER_MONITOR
WHERE TORNADO =0 AND YEAR = 2021
GROUP BY YEAR, MONTH
"""
pd.read_sql(sql,conn)

YEAR  MONTH  TOTALTORNADO_RAIN  TOTAL_NON_TORNADO_RAIN
0  2021      1             316.27                  316.27
1  2021      2             122.85                  122.85
2  2021      3             104.11                  104.11
3  2021      4             143.92                  143.92
4  2021      5             138.36                  138.36

In [29]:
sql = """
--Solution AVEC CASE
SELECT CAST(strftime('%Y', REPORT_DATE) AS INTEGER) AS YEAR,
       CAST(strftime('%m', REPORT_DATE) AS INTEGER ) AS MONTH,
       SUM( CASE WHEN TORNADO = 1 THEN RAIN ELSE 0 END ) AS TOTALTORNADO_RAIN,
       SUM( CASE WHEN TORNADO = 0 THEN RAIN ELSE 0 END ) AS TOTAL_NON_TORNADO_RAIN
FROM WEATHER_MONITOR
WHERE YEAR = 2021
GROUP BY YEAR, MONTH
"""
pd.read_sql(sql,conn)

YEAR  MONTH  TOTALTORNADO_RAIN  TOTAL_NON_TORNADO_RAIN
0  2021      1               0.00                  316.27
1  2021      2              15.22                  122.85
2  2021      3              24.92                  104.11
3  2021      4               9.87                  143.92
4  2021      5              19.88                  138.36

Cependant, un 0 pour la condition fausse peut être problématique pour d'autres opérations d'agrégation comme MIN, MAX et AVG et COUNT car cela affectera ces calculs contrairement à SUM. Vous pouvez plutôt utiliser NULL car il sera ignoré par tous les opérateurs d'agrégation, y compris SUM.

In [30]:
sql = """
--Solution AVEC CASE et NULL
SELECT CAST(strftime('%Y', REPORT_DATE) AS INTEGER) AS YEAR,
       CAST(strftime('%m', REPORT_DATE) AS INTEGER ) AS MONTH,
       SUM( CASE WHEN TORNADO = 1 THEN RAIN ELSE NULL END ) AS TOTALTORNADO_RAIN,
       SUM( CASE WHEN TORNADO = 0 THEN RAIN ELSE NULL END ) AS TOTAL_NON_TORNADO_RAIN
FROM WEATHER_MONITOR
WHERE YEAR = 2021
GROUP BY YEAR, MONTH
"""
pd.read_sql(sql,conn)

YEAR  MONTH  TOTALTORNADO_RAIN  TOTAL_NON_TORNADO_RAIN
0  2021      1                NaN                  316.27
1  2021      2              15.22                  122.85
2  2021      3              24.92                  104.11
3  2021      4               9.87                  143.92
4  2021      5              19.88                  138.36

```sql
--T sql
SELECT
    YEAR(REPORT_DATE) AS YEAR,
    MONTH(REPORT_DATE) AS MONTH,
    SUM(CASE WHEN TORNADO = 1 THEN RAIN ELSE 0 END) AS TOTALTORNADO_RAIN,
    SUM(CASE WHEN TORNADO = 0 THEN RAIN ELSE 0 END) AS TOTAL_NON_TORNADO_RAIN
FROM
    WEATHER_MONITOR
WHERE
    YEAR(REPORT_DATE) = 2021
GROUP BY
    YEAR(REPORT_DATE), MONTH(REPORT_DATE);

```

# EXERCICE
Pour chaque LOCATION_ID, calculez la pluie totale de l'année précédente **PY_RAIN** (2020) et la pluie totale de l'année en cours **CY_RAIN** (2021).


* Ceci est le résultat pour les 5 premières lignes

| | LOCATION_ID | CY_RAIN | PY_RAIN |
|---:|--------------:|----------:|----------:|
|  0 |             0 |     18.99 |     24.72 |
|  1 |             1 |     22.51 |     45.02 |
|  2 |             2 |     30.72 |     16.04 |
|  3 |             3 |     15.57 |     22.8  |
|  4 |             4 |     30.17 |     30.15 |
|  5 |             5 |     40.71 |      7.92 |

In [31]:
# by SYLVIE
sql = """
SELECT
  LOCATION_ID,
  ROUND(SUM(CASE WHEN strftime('%Y', REPORT_DATE) = '2021' THEN RAIN ELSE NULL END), 2) AS CY_RAIN,
  ROUND(SUM(CASE WHEN strftime('%Y', REPORT_DATE) = '2020' THEN RAIN ELSE NULL END), 2) AS PY_RAIN
FROM WEATHER_MONITOR
GROUP BY LOCATION_ID
ORDER BY LOCATION_ID
LIMIT 6"""
pd.read_sql(sql,conn)

LOCATION_ID  CY_RAIN  PY_RAIN
0            0    18.99    24.72
1            1    22.51    45.02
2            2    30.72    16.04
3            3    15.57    22.80
4            4    30.17    30.15
5            5    40.71     7.92

In [32]:
# by KENZA
sql = """
SELECT LOCATION_ID,
       SUM( CASE WHEN CAST(strftime('%Y', REPORT_DATE) AS INTEGER) = 2021 THEN RAIN ELSE null END ) AS PY_RAIN,
       SUM( CASE WHEN CAST(strftime('%Y', REPORT_DATE) AS INTEGER) = 2020 THEN RAIN ELSE null END ) AS CY_RAIN
FROM WEATHER_MONITOR
GROUP BY LOCATION_ID
LIMIT 6
"""
pd.read_sql(sql,conn)

LOCATION_ID  PY_RAIN  CY_RAIN
0            0    18.99    24.72
1            1    22.51    45.02
2            2    30.72    16.04
3            3    15.57    22.80
4            4    30.17    30.15
5            5    40.71     7.92